src/lib.rs

```diff
@@ -0,0 +1,550 @@
+    COMPUTE_HEAVY_FUTURE_EXECUTOR_STRATEGY
+        .set(strategy)
+        .unwrap_or_else(|_| panic!("Compute heavy future executor can only be initialized once"))
+}
+
+/// The stored strategy used to spawn compute-heavy futures.
+static COMPUTE_HEAVY_FUTURE_EXECUTOR_STRATEGY: OnceLock<ExecutorStrategy> = OnceLock::new();
+
+trait ComputeHeavyFutureExecutor {
+    /// Accepts a future and returns its result
+    async fn execute<F, O>(&self, fut: F) -> Result<O, Error>
+    where
+        F: Future<Output = O> + Send + 'static,
+        O: Send + 'static;
+}
+
+enum ExecutorStrategy {
+    /// A non-op strategy that awaits in the current context
+    CurrentContext(CurrentContextExecutor),
+    /// User-provided closure
+    CustomExecutor(CustomExecutor),
+    /// tokio task::spawn_blocking
+    #[cfg(compute_heavy_executor_tokio)]
+    SpawnBlocking(SpawnBlockingExecutor),
+    /// tokio task::block_in_place
+    #[cfg(compute_heavy_executor_tokio)]
+    BlockInPlace(BlockInPlaceExecutor),
+    #[cfg(compute_heavy_executor_tokio)]
+    /// Spin up a second, lower-priority tokio runtime
+    /// that communicates via channels
+    SecondaryTokioRuntime(SecondaryTokioRuntimeExecutor),
+}
+
+<<<<<<< Updated upstream
+impl ExecutorStrategy {}
+
+=======
+>>>>>>> Stashed changes
+/// Spawn a future to the configured compute-heavy executor and wait on its output.
+/// 
+/// # Strategy selection
+/// 
+/// If no strategy is configured, this library will fall back to the following defaults:
+/// - no tokio cfg flag - current context
+/// - tokio, in multi-threaded runtime flavor - block in place
+/// - tokio, all other flavors - spawn blocking
+/// 
+/// You can override these defaults by initializing a strategy:
+/// - [`initialize_current_context_strategy`],
+/// - [`initialize_spawn_blocking_strategy`]
+/// - [`initialize_block_in_place_strategy`]
+/// - [`initialize_secondary_tokio_runtime_strategy`]
+/// - [`initialize_secondary_tokio_runtime_strategy_and_config`]
+/// 
+/// # Example 
+/// 
+/// ```
+/// # async fn run() {
+/// use compute_heavy_future_executor::spawn_compute_heavy_future;
+/// 
+/// let future = async { 
+///     std::thread::sleep(std::time::Duration::from_millis(50));
+///     5
+///  }; 
+/// 
+/// let res = spawn_compute_heavy_future(future).await.unwrap();
+/// assert_eq!(res, 5);
+/// # }
+/// 
+/// ```
+/// 
+pub async fn spawn_compute_heavy_future<F, R>(fut: F) -> Result<R, Error>
+where
+    F: Future<Output = R> + Send + 'static,
+    R: Send + 'static,
+{
+    let executor = COMPUTE_HEAVY_FUTURE_EXECUTOR_STRATEGY
+        .get_or_init(|| {
+                #[cfg(compute_heavy_executor_tokio)]
+                match tokio::runtime::Handle::current().runtime_flavor() {
+                    tokio::runtime::RuntimeFlavor::MultiThread => {
+                        log::info!("spawn_compute_heavy_future called without setting an explicit executor strategy, \
```

Comment:
The `log` should probably be `trace` in production to avoid creating too much log spam.